### PR TITLE
Add `strict-type-predicates` rule

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [X] Documentation update
- [X] Enable CircleCI for your fork (https://circleci.com/add-projects)

#### What changes did you make?

Added the `strict-type-predicates` rule, which warns for `typeof x === "string"` where `x` is always (or never) a string, and for `x === undefined` where `x` is always (or never) `undefined`.

#### Is there anything you'd like reviewers to focus on?

Can't do `instanceof` checks yet as the type checker does not expose assignability checks. If it did, the `getTypePredicateForKind` function could be removed.

### CODE REVIEW

File: test/rules/strict-type-predicates/test.ts.lint [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    get<number | undefined>() === null;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    get<number | null>() === null;
+
+    get<number | null>() === undefined;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
+
+    get<number | undefined>() === undefined;
+    get<string | undefined>() == null;
```
Comment (ajafff) [inline test/rules/strict-type-predicates/test.ts.lint]: Could this rule also check for non strict comparisons to `null` or `undefined` if only one of them is possible?
Example:
```typescript
get<string|undefined>() == null;
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [use === undefined instead]

get<string|undefined|null>() == null; // still allowed
```

File: test/rules/strict-type-predicates/test.ts.lint [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    get<string | null>() == undefined;
+    get<string | null>() == null;
+    get<string | undefined>() == undefined;
```
Comment (ajafff) [inline test/rules/strict-type-predicates/test.ts.lint]: here the rule could suggest to use strict equality comparison

Comment (ajafff) [inline test/rules/strict-type-predicates/test.ts.lint]: Please add a test for the following:

```typescript
get<null|undefined> == null;
~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [T]
get<null|undefined> != undefined;
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [F]
```

**Comment by andy-hanson** [comment]
Would this rule make `typeof-compare` unnecessary?

**Comment by mohsen1** [comment]
@andy-hanson if it catches typos like `typeof foo === 'Object'` it should...

**Comment by nchen63** [comment]
sorry, can you merge with master?

**Pull request merged**